### PR TITLE
introduce WriteSynchronizer

### PR DESCRIPTION
**Pull request by kevodwyer**

### CODE REVIEW

File: src/peergos/shared/user/UserContext.java [hunk 1]
```diff
@@ -349,17 +350,18 @@ public boolean isJavascript() {
...
         progressCallback.accept("Retrieving Friends");
-        return userData.getValue()
-                .thenCompose(wd -> createFileTree(entrie, username, network, crypto.random, crypto.hasher)
+        return writeSynchronizer.getCurrentWriterData(signer.publicKeyHash, signer.publicKeyHash,
+                wd -> createFileTree(entrie, username, network, crypto.random, crypto.hasher)
                         .thenCompose(root -> {
                             this.entrie = root;
                             return getByPath("/" + username + "/" + "shared")
-                                    .thenCompose(sharedOpt -> {
+                                    .thenApply(sharedOpt -> {
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: this should still be thenCompose I believe

File: src/peergos/shared/user/UserContext.java [hunk 2]
```diff
@@ -349,17 +350,18 @@ public boolean isJavascript() {
...
                                         if (!sharedOpt.isPresent())
                                             throw new IllegalStateException("Couldn't find shared folder!");
                                         return buildSharedWithCache(sharedOpt.get(), this::getUserRoot);
-                                    }).thenApply(res -> this);
-                        }));
+                                    });
+                        }).thenApply(b -> wd)
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: don't need this thenApply

File: src/peergos/shared/user/UserContext.java [hunk 3]
```diff
@@ -518,30 +520,33 @@ public boolean isJavascript() {
...
-                                        return network.coreNode.updateChain(username, claimChain)
-                                                .thenCompose(updatedChain -> {
-                                                    if (!updatedChain)
-                                                        throw new IllegalStateException("Couldn't register new public keys during password change!");
-
-                                                    return UserContext.ensureSignedUp(username, newPassword, network, crypto);
-                                                });
-                                    });
-                                }));
+                                ).thenCompose(newSignerHash ->
```
Comment (ianopolous) [inline src/peergos/shared/user/UserContext.java]: if only indentation has changed here can we revert it?

File: src/peergos/shared/user/UserContext.java (hunk 1 above)
Comment (kevodwyer) [reply src/peergos/shared/user/UserContext.java]: done

File: src/peergos/shared/user/UserContext.java (hunk 3 above)
Comment (kevodwyer) [reply src/peergos/shared/user/UserContext.java]: nope, the code structure has changed.

File: src/peergos/shared/user/UserContext.java (hunk 2 above)
Comment (kevodwyer) [reply src/peergos/shared/user/UserContext.java]: i think it is necessary...

**Pull request merged**